app/views/layouts/_head.html.erb

```diff
@@ -2,6 +2,7 @@
   <meta http-equiv="X-UA-Compatible" content="IE=edge" />
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <%= javascript_include_tag "es6" unless browser.es6? %>
+  <%= javascript_include_tag "turbo", :type => "module" %>
```

Comment:
What was the thinking behind making this a separate include rather than adding it to the application bundle? or is it necessary in order to include it as a module? I confess to not understanding javascript modules and how they interact with other sorts of javascript and rails assets very well...

Reply:
This is the only way I was able to import Turbo (which comes bundled in ESM format in the `turbo-rails`-gem).

Trying to import it inside the `application.js` with Sprockets raises errors:

```diff
diff --git a/app/assets/javascripts/application.js b/app/assets/javascripts/application.js
index c71c0652d..25ca33d34 100644
--- a/app/assets/javascripts/application.js
+++ b/app/assets/javascripts/application.js
@@ -18,6 +18,7 @@
 //= require qs/dist/qs
 //= require bs-custom-file-input
 //= require bs-custom-file-input-init
+//= require turbo
 
 /*
  * Called as the user scrolls/zooms around to manipulate hrefs of the
diff --git a/app/views/layouts/_head.html.erb b/app/views/layouts/_head.html.erb
index 3c691612a..0ee5bf9c0 100644
--- a/app/views/layouts/_head.html.erb
+++ b/app/views/layouts/_head.html.erb
@@ -2,7 +2,6 @@
   <meta http-equiv="X-UA-Compatible" content="IE=edge" />
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <%= javascript_include_tag "es6" unless browser.es6? %>
-  <%= javascript_include_tag "turbo", :type => "module" %>
   <%= javascript_include_tag "application" %>
   <%= javascript_include_tag "i18n/#{I18n.locale}" %>
   <%= stylesheet_link_tag "screen-#{dir}", :media => "screen" %>
```

![SCR-20240320-rxmw](https://github.com/openstreetmap/openstreetmap-website/assets/28648/914c1b79-6308-47c5-a23b-a2f11c48e60d)

Reply:
Since modules are deferred, you had to defer the init code:

https://github.com/openstreetmap/openstreetmap-website/pull/4562/files#diff-0e145ba000e96b1a45d230275e350e4e8514efb4db7dbcfa228fe1003896c9b4R72-R74

Is there a way for Sprockets to translate a module into non-module?

Reply:
I've been looking and I can't see any obvious solution so it's probably not worth spending any more time on it when we have something that works - the sprockets asset pipeline is kind of deprecated now anyway.

There is a separate problem that I have discovered though, which is that turbo is trying to inject CSS for the progress bar and failing as our CSP rules prohibit inline CSS rules. It has code to work with rails builting security policy support but we're still using secure_headers so https://github.com/openstreetmap/openstreetmap-website/blob/8da80ff471f7d04984ff5abab0c5fe32920330ba/app/views/layouts/_head.html.erb#L14 needs to be removed and replaced with this:

```
<meta name="csp-nonce" content="<%= content_security_policy_style_nonce %>" />
```

which sets the CSP nonce for turbo using secure_headers instead.

Reply:
> […] which is that turbo is trying to inject CSS for the progress bar and failing as our CSP rules prohibit inline CSS styles

I haven't noticed that. Did this come up as a warning/error in the JS console or how did you notice?

Reply:
Yes you should get a warning in the console.

Reply:
Ah you'll need to have `csp_enforce:true` in `config/settings.local.yml` as well.

Reply:
ah! Now I can see it.

![SCR-20240321-hvcz](https://github.com/openstreetmap/openstreetmap-website/assets/28648/5a14104c-9e53-4589-a865-01af6a3179ce)

some questions :)

I see that the file is tracked in git. Is that `settings.yml` loaded in Production and `enforce_csp: false`?

---

I'll have a closer look at the error. My CSP knowledge is still limited. I'll have to do a bit of homework first.

Reply:
Sorry yes that file is for local overrides of the defaults. There's a live version with my [fix](https://github.com/tomhughes/openstreetmap-website/commit/1b9d82d66aba56453ed75070637945967a30e4d5) running at https://tomh.apis.dev.openstreetmap.org/.

Reply:
> which sets the CSP nonce for turbo using secure_headers instead.

ok. I had to read up a bit on the `secure_headers`-gem CSP implementation and CSP generally.

I understand the the `csp_meta_tag` currently does nothing since its provided by Rails, which builds on Rails CSP configuration (currently nothing is configured in `config/initializers/content_security_policy.rb`).

`content_security_policy_style_nonce` is provided by `secure_headers` (here https://github.com/github/secure_headers/blob/7a23cb6b350b024a786e163e81c902552b9c484f/lib/secure_headers/view_helper.rb#L83-L85) and therefore the correct way to safelist certain inline styles.

I'll adjust the PR to reflect this. Thanks @tomhughes for pointing this out and please correct anything or add anything you think one should know.

As I've mentioned my understanding of CSP is still limited and I have not practical experiences yet.

Reply:
Actually 🤔 thinking about it for another second: The change of replacing `csp_meta_tag` with `<meta name="csp-nonce" content="<%= content_security_policy_style_nonce %>" />` is only indirectly related to this PR, no?

So to keep the PR and its changeset smaller I'd like to move this change into a separate PR. Does that sounds right? @tomhughes

Reply:
Well we haven't used nonces before so it hasn't been an issue before now but sure that can be a separate pull request. I think I have a branch somewhere where I started on moving to the rails builtin policy stuff so I should probably resurrect that at some point...

Reply:
https://github.com/openstreetmap/openstreetmap-website/pull/4606 addressed the CSP violation